File: src/web-ui/src/web-ui/src/infrastructure/api/tupai/automationPrompt.ts
// Track F — interactive prompt API.
//
// Mirrors `src-tauri/src/pc_automation/skill/types.rs` (camelCase
// wire shape) and bridges the front-end to the two Tauri commands
// (`automation_answer_prompt` / `automation_cancel_prompt`) plus the
// `automation:ask_user` event the executor emits when a `SkillStep`
// carries an `InteractionPrompt`.
//
//   * `onAskUser(handler)`   — subscribe to `automation:ask_user`;
//                               the handler renders the modal.
//   * `answerPrompt(...)`    — deliver the user's answer back to the
//                               blocked executor via the
//                               `automation_answer_prompt` command.
//   * `cancelPrompt(...)`    — signal dismissal via
//                               `automation_cancel_prompt`; the
//                               executor then falls back to
//                               `default_value`.
//
// `invoke` is the tupai wrapper (`./invoke`) that no-ops in non-Tauri
// runtimes (pnpm dev / web preview) so this module is safe to import
// from anywhere. `@tauri-apps/api/event` is already a dependency
// (used by 30+ files across the app).

import { invoke } from './invoke';
import { listen, type UnlistenFn } from '@tauri-apps/api/event';

export interface PromptChoice {
  id: string;
  label: string;
}

export type PromptInputType = 'text' | 'choice' | 'multichoice' | 'confirm';

export interface InteractionPrompt {
  promptId: string;
  question: string;
  inputType: PromptInputType;
  choices: PromptChoice[];
  bindToVar: string;
  defaultValue?: unknown;
  timeoutMs: number;
}

export interface AskUserPayload {
  correlationId: string;
  skillId: string;
  stepId: string;
  prompt: InteractionPrompt;
}

export interface PromptAnswer {
  correlationId: string;
  value: unknown;
  cancelled: boolean;
}

/**
 * Deliver the user's answer to a pending `automation:ask_user`
 * prompt. `value` is the typed string for `text`, the `choice.id`
 * for `choice`, an array of selected choice ids for `multichoice`,
 * and `"true"` / `"false"` for `confirm`.
 */
export async function answerPrompt(
  correlationId: string,
  value: unknown,
  cancelled = false,
): Promise<void> {
  return invoke<void>('automation_answer_prompt', {
    answer: { correlationId, value, cancelled },
  });
}

/** Signal that the user dismissed the prompt without answering. */
export async function cancelPrompt(correlationId: string): Promise<void> {
  return invoke<void>('automation_cancel_prompt', { correlationId });
}

/**
 * Subscribe to `automation:ask_user` events. Returns an `UnlistenFn`
 * that the caller should invoke on teardown. The handler receives
 * the decoded `AskUserPayload` (the event's `payload` field).
 */
export function onAskUser(handler: (payload: AskUserPayload) => void): Promise<UnlistenFn> {
  return listen<AskUserPayload>('automation:ask_user', (e) => handler(e.payload));
}

File: src-tauri/src/commands/automation_prompt.rs
// Copyright (c) 2026 AIMarketing
//
// Track F — Tauri command surface for interactive prompts.
//
// The executor emits `automation:ask_user` and blocks on
// `prompt_registry::register(correlation_id)`. The front-end, after
// rendering the prompt modal, calls back into one of these two
// commands to deliver the answer (or signal cancellation):
//
//   * `automation_answer_prompt(answer)` — fulfills the oneshot the
//     executor is awaiting. `answer.value` carries the user input
//     (text / choice.id / "true"|"false"); `answer.cancelled`
//     flags a dismiss.
//   * `automation_cancel_prompt(correlation_id)` — drops the
//     pending sender so the executor's `recv()` returns
//     `Err(RecvError)`; its timeout match then falls back to
//     `default_value`.
//
// NOTE: These two commands are registered in `lib.rs`'s
// `invoke_handler!`, so the front-end can deliver answers /
// cancellations directly via Tauri IPC.

use tauri::AppHandle;

use crate::pc_automation::executor::prompt_registry;
use crate::pc_automation::skill::types::PromptAnswer;

/// Deliver the user's answer to a pending `automation:ask_user`
/// prompt. The `correlation_id` inside `answer` must match the id
/// the executor emitted. Returns `Err` if no pending prompt exists
/// for that id (already timed out / cancelled / unknown).
#[tauri::command]
pub async fn automation_answer_prompt(answer: PromptAnswer, _app: AppHandle) -> Result<(), String> {
    prompt_registry::deliver(answer)
}

/// Cancel a pending prompt by `correlation_id`. Best-effort: no-op
/// if the prompt was never registered or already fulfilled. The
/// executor's await will then fall through to `default_value`.
#[tauri::command]
pub fn automation_cancel_prompt(correlation_id: String, _app: AppHandle) -> Result<(), String> {
    prompt_registry::cancel(&correlation_id);
    Ok(())
}
